src/osd/OSDMap.cc

```diff
@@ -3809,21 +3809,30 @@ string OSDMap::get_flag_string() const
 
 void OSDMap::print_pools(ostream& out) const
 {
-  for (const auto &pool : pools) {
+  for (const auto &[pid, pdata] : pools) {
     std::string name("<unknown>");
-    const auto &pni = pool_name.find(pool.first);
+    const auto &pni = pool_name.find(pid);
     if (pni != pool_name.end())
       name = pni->second;
-    out << "pool " << pool.first
+    char wl_score_str[32];
+    if (pdata.is_replicated())
+      snprintf (wl_score_str, sizeof(wl_score_str),
+		" workload_balance_score %.2f",
+		calc_wl_balance_score(g_ceph_context, pid));
```

Comment:
```
FAILED: src/crimson/CMakeFiles/crimson-common.dir/__/osd/OSDMap.cc.o 
/usr/bin/ccache /usr/bin/clang++-14 -DBOOST_ASIO_DISABLE_CONCEPTS -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DCEPH_INSTALL_DATADIR=\"/usr/local/share/ceph\" -DCEPH_INSTALL_FULL_PKGLIBDIR=\"/usr/local/lib/ceph\" -DCMAKE_INSTALL_LIBDIR=\"lib\" -DHAVE_CONFIG_H -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_TYPE_ERASE_MORE -DWITH_SEASTAR=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/src -I/home/jenkins-build/build/workspace/ceph-pull-requests/src/seastar/include -I/home/jenkins-build/build/workspace/ceph-pull-requests/build/src/seastar/gen/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/build/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/xxHash -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/rapidjson/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/api/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /home/jenkins-build/build/workspace/ceph-pull-requests/src/jaegertracing/opentelemetry-cpp/sdk/include -g -Werror -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -Wno-non-virtual-dtor -std=gnu++17 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Wno-error=unused-result "-Wno-error=#warnings" -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -std=c++20 -MD -MT src/crimson/CMakeFiles/crimson-common.dir/__/osd/OSDMap.cc.o -MF src/crimson/CMakeFiles/crimson-common.dir/__/osd/OSDMap.cc.o.d -o src/crimson/CMakeFiles/crimson-common.dir/__/osd/OSDMap.cc.o -c /home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/OSDMap.cc
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/OSDMap.cc:3959:25: error: use of undeclared identifier 'g_ceph_context'
                calc_wl_balance_score(g_ceph_context, pid));
                                      ^
1 error generated.
```

The compiler testifies `g_ceph_context` is undeclared here. This may happen in crimson as we have `WITH_SEASTAR` in multiple places, also around includes. How about including `global/global_context.h`?

To replicate locally you would need to `install-deps.sh --crimson` and also `do_cmake.sh -DWITH_SEASTAR=ON`.

Reply:
@rzarzynski 
For some reasons `install-deps.sh --crimson` did not do anything for me, I ran `do-cmake.sh` several times and installed the missing packages that it found, but now I get the following errors which I don't know how to solve 
`CMake Error at src/seastar/CMakeLists.txt:754 (message):`
`  Sanitizers not found!`

Reply:
@JoshSalomon Radek has raised this PR to help with this: https://github.com/ceph/ceph/pull/47697